README.md
# Amaze UI Touch

基于 React.js 的移动端 Web UI 组件库。

[![NPM version](https://img.shields.io/npm/v/amazeui-touch.svg?style=flat-square)](https://www.npmjs.com/package/amazeui-touch)
[![Build Status](https://img.shields.io/travis/amazeui/amazeui-touch.svg?style=flat-square)](https://travis-ci.org/amazeui/amazeui-touch)
[![Dependency Status](https://img.shields.io/david/amazeui/amazeui-touch.svg?style=flat-square)](https://david-dm.org/amazeui/amazeui-touch)
[![devDependency Status](https://img.shields.io/david/dev/amazeui/amazeui-touch.svg?style=flat-square)](https://david-dm.org/amazeui/amazeui-touch#info=devDependencies)

### [入门套件](https://github.com/amazeui/amt-starter-kit)

## 简介

### 专属于移动

Amaze UI Touch 专为移动打造，在技术实现、交互设计上只考虑主流移动设备，保证代码轻、性能高。

### 专注于 UI

只提供 UI 组件（View），对配套技术不做限定，方便用户与现有技术栈快速整合，降低使用成本。

### 采用 Flexbox

使用 Flexbox 技术，灵活自如地对齐、收缩、扩展元素，轻松搞定移动页面布局。

### 基于 React.js

基于风靡社区的 React.js 封装组件，沿袭高性能、可复用、易扩展、一处学习多端编写特性。



## 安装使用

1. 从 npm 安装：

    ``` bash
    npm install --save-dev amazeui-touch
    ```

2. 使用 组件：

    ``` javascript
    class App extends React.Component {
      constructor(props) {
        super(props);
        this.state = {
          said: false,
        };

        this.handleClick = this.handleClick.bind(this);
      }

      handleClick() {
        this.setState({
          said: true,
        });
      }

      renderHello() {
        return this.state.said ? (
          <p>Hello World! Welcome to Amaze UI Touch.</p>
        ) : null;
      }

      render() {
        const said = this.state.said;
        const text = said ? 'Said :(' : 'Say hello :)';

        return (
          <div>
            <Button
              amStyle="primary"
              disabled={said}
              onClick={this.handleClick}
            >
              {text}
            </Button>
            {this.renderHello()}
          </div>
        );
      }
    }

    ReactDOM.render(<App />, document.getElementById('root'));
    ```

ES2015/JSX 编译参见 [Amaze UI Touch Starter Kit](https://github.com/amazeui/amt-starter-kit)。



## 文档及演示

1. 全局安装 gulp.js：

   ``` bash
   npm install -g gulp
   ```


1. 克隆源代码并安装依赖：

   ``` bash
   git clone https://github.com/amazeui/amazeui-touch.git
   ```

   在源码目录下执行：

   ``` bash
   npm install
   ```

2. 执行 gulp：

   ``` bash
   gulp
   ```

   相关文件构建完成后会自动打开浏览器，可查看文档及组件演示。

   ​

### [反馈建议](https://github.com/amazeui/amazeui-touch/issues)



## License

MIT © 2015 - 2016 AllMobilize Inc.


CHANGELOG.md
## Change Log

### v1.0.0-beta.2 (2016/05/24 15:44 +08:00)

- `Improved` [e1b0e29](https://github.com/amazeui/amazeui-touch/commit/e1b0e29ac65e2b9c6fc44f6567128af0361798f6) #50 解决 `context` 在 OffCanvas 和 Popover 中丢失问题
- `New` [3aa23d4](https://github.com/amazeui/amazeui-touch/commit/3aa23d4c23437a5d5cf7bdbef83bd8927f0b398d) Android UC 浏览器在 `<html>` 上添加 `.ua-stupid-uc` 类名，以应用相关回退样式
- `Changed` [1b988fd](https://github.com/amazeui/amazeui-touch/commit/1b988fd77f8a3625a8e36d1c30afada871482dca) React 相关包移到 `peerDependencies` 字段，防止打包多个版本 React 报错问题
- `Fixed` [6977df2](https://github.com/amazeui/amazeui-touch/commit/6977df245e3f45dc068fadbab217bcd3d5d9b83e) 修复 Slider 在以 Android UC 为代表的浏览器中 `swipe` 手势无效问题
- `Improved` [d72e6ce](https://github.com/amazeui/amazeui-touch/commit/d72e6cee6622f92683c1c2b591ee37c8a9a63936) 网格系统兼容 Android UC 为代表的只支持 2009 版 flexbox 规范的浏览器
- `Fixed` [814f8dc](https://github.com/amazeui/amazeui-touch/commit/814f8dc206b40a75cfab5d08792bb3c53e2a0f12)  #31 文档在 Android UC 等浏览器中无法滚动问题
- `Improved` [6769e0f](https://github.com/amazeui/amazeui-touch/commit/6769e0fed9f0c5b7a175163a99344c2cf2e17dfb) #44 Switch 增加 `value` `disabled` `getValue` API
- `Changed` [eccf8e5](https://github.com/amazeui/amazeui-touch/commit/eccf8e5d6da12cd287c33559cab2a59d695616b6)
减少 NavBar 导航条目数据层级
- `Improved` [470741e](https://github.com/amazeui/amazeui-touch/commit/470741ee254d4d6898d146eee77614969c04652a) 处理 NavBar 在以 Android UC 为代表的浏览器中的兼容性 (@minwe)
- `New` [fcf9893](https://github.com/amazeui/amazeui-touch/commit/fcf98932c08b2ae67fbfd90f1844f9e070d09b73) 增加 `margin` `padding` 及文字颜色相关辅助类名
- `Improved` [f4f9a29](https://github.com/amazeui/amazeui-touch/commit/f4f9a29bdbb49ec8c763414bb745850ce8f697f7) #23 改进 Container 在 iOS 中流畅度
- `Improved` [0f680f8](https://github.com/amazeui/amazeui-touch/commit/0f680f823648eb63528195f1907c1e4e1830381a) 重构 Notification
- `Changed` [84c682e](https://github.com/amazeui/amazeui-touch/commit/84c682e9a606e9f8cc1d697e733d19e4721e596d) Modal `onRequestClose` 属性重命名为 `onDismiss`
- `Improved` [ed81459](https://github.com/amazeui/amazeui-touch/commit/ed81459aba1180aaf36ffa08080231919379cf0a) 重构 Modal，增强可定制性
- `Improved` [a79fe65](https://github.com/amazeui/amazeui-touch/commit/a79fe65c86e31153a201cce5033e1e7a99daf317) #40 调整 NavBar API，可以在链接上放置 OffCanvasTrigger


### v1.0.0-beta.1 (2016/04/25)

- `Changed` [93fac1f](https://github.com/amazeui/amazeui-touch/commit/93fac1f8a17a1f9d61af3e401bb759f2f2467a3b) #39 **API 变更**：`onSelect` 与 React 事件 API 冲突，重命名为 `onAction`，涉及的组件包括：
  - Modal
  - NavBar
  - Slider
  - Tabs
  - TabBar
- `Fixed` [89bc1ff](https://github.com/amazeui/amazeui-touch/commit/89bc1ff022f3b9074409c33f09c24af3c471349f) 修复 React 15.x 读取 `key` 警告问题
- `Changed` [721a2a7](https://github.com/amazeui/amazeui-touch/commit/721a2a72a63c20200b494d76655f81aca80914cd) #36 `react` 和 `react-dom` 移到 `peerDependencies`，解决使用时打包多个版本的问题
- `Fixed` [3d6bbdc](https://github.com/amazeui/amazeui-touch/commit/3d6bbdc819175a0cf6b31185b2a7b3421e447a8c) 修复 flexbox 在部分安卓 UA 上的显示问题（Thx @leakl），至此 flexbox 兼容性问题主要剩下 NavBar 和 Grid
- `Improved` [082c00f](https://github.com/amazeui/amazeui-touch/commit/082c00f2ad7d1e61c914238cc5eb326f73cb672e) 运行环境判断，为后端渲染做准备
- `Fixed` [4903567](https://github.com/amazeui/amazeui-touch/commit/49035679e1e3c55e043eeae02df1dd62be8c71c9) #27 修复 Object.assign 没有转换问题
- `Fixed` [dcdf80a](https://github.com/amazeui/amazeui-touch/commit/dcdf80a3dcbf1611c5422d4ded8b39de7781f9a0) 修复 Modal `open()` 方法判断逻辑始终为 `true` 的错误

### v0.1.0-beta2 (2015/11/17)

- `Improved` [81c4c1a](https://github.com/amazeui/amazeui-touch/commit/81c4c1a23fa5fca4a3352e3a8711a79fa04d7b3b) 移除 `input[type="range"]` 在 Safari 中的黑色边框
- `Improved`  [e39f2e0](https://github.com/amazeui/amazeui-touch/commit/e39f2e023ce9e7997adcafbb96e0cfbc0274ea97) #6 调整 View 样式
- `Improved` #11 优化 Modal、Notification 等组件在 iOS 9 Safari 中入场动画效果
- `Improved` [6cdf4ec](https://github.com/amazeui/amazeui-touch/commit/6cdf4ec61c357b0471837a49db78ba1d66f5b564) #5 修复 Slider 阻止垂直触控默认滚动行为问题
- `Fixed` [33305f4](https://github.com/amazeui/amazeui-touch/commit/33305f49405b09ec578fdf530a54012e1fd9bf43)  #7 修复 Modal 遮罩层显示问题（via [519ae20](https://github.com/amazeui/amazeui-touch/commit/519ae20c8646252c06b819c538c74395a4e47b22) ）
- `Changed` [44bfec0](https://github.com/amazeui/amazeui-touch/commit/44bfec03b4fbee022914476abb88b3f090f8d093) #10 替换 Loader 样式，移除 SVG
- `Improved` [ea51873](https://github.com/amazeui/amazeui-touch/commit/ea51873871d70e131b1d307a1a4236d89e37774e) Button 添加 `:hover` 样式
- `Improved` [338dca4](https://github.com/amazeui/amazeui-touch/commit/338dca48a0fec187ebbcd6215853f27942c6ab99) 官网添加 kitchen-sink 二维码


gulpfile.babel.js
/**
 * Amaze UI Touch Building Tasks
 *
 * @author Minwe <minwe@yunshipei.com>
 */

import path from 'path';
import gulp from 'gulp';
import gulpLoadPlugins from 'gulp-load-plugins';
import del from 'del';
import runSequence from 'run-sequence';
import webpack from 'webpack-stream';
import named from 'vinyl-named';
import assign from 'object-assign';
import webpackConfig from './webpack.config';
import browserify from 'browserify';
import watchify from 'watchify';
import source from 'vinyl-source-stream';
import buffer from 'vinyl-buffer';
import merge from 'merge-stream';
import markedify from 'markedify';
import BS from 'browser-sync';
import pkg from './package.json';
import getMarked from './docs/_utils/getMarked';

import through from 'through2';

const ENV = process.env.NODE_ENV;
const $ = gulpLoadPlugins();
const isProduction = ENV === 'production' || ENV === 'travisci';
const banner = `/** ${pkg.title} v${pkg.version} | by Cloud Team
  * (c) ${$.util.date(Date.now(), 'UTC:yyyy')} Thinker, Inc., Licensed under ${pkg.license}
  * ${$.util.date(Date.now(), 'isoDateTime')}
  */
  `;

const paths = {
  style: [
    'src/scss/amazeui.touch.scss'
  ],
  scssModules: 'src/scss/**/*.scss',
  fonts: 'src/fonts/*',
  // jsEntry: ['src/js/index.js'],
  jsEntry: 'src/js/index.js',
  dist: 'dist',
};

// const docsDir = 'docs/_app';
// const appPaths = {
//   js: `${docsDir}/js/app.js`,
//   styleDir: `${docsDir}/style`,
//   style: `${docsDir}/style/app.scss`,
//   dist: 'www',
//
//   ksEntry: 'kitchen-sink/app.js',
//   ksIndex: 'kitchen-sink/index.html',
//   ksDist: 'www/kitchen-sink',
//
//   appEntry: 'app/js/app.js',
//   appIndex: 'app/index.html',
//   appDist: 'www/app'
// };
var appDir = 'docs/_app';
var appPaths;
var resetPaths = function(dir, entryJs){
  entryJs = entryJs || 'app.js';
  appDir = dir || 'docs/_app';
  appPaths = {
    quoteSrc: `src`,
    quoteStyles: `src/styles/quote.scss`,
    quotePage: `${appDir}/index.html`,
    imgs: `${appDir}/i/*`,
    js: `${appDir}/js/app.js`,
    styleDir: `${appDir}/style`,
    style: [`${paths.style}`, `${appDir}/style/app.scss`],
    dist: dir ? `www/${appDir}` : 'www',

    appEntry: `${appDir}/js/${entryJs}`,
    appIndex: `${appDir}/index.html`,
    appDist: `www/${appDir}`
  }
  return appPaths;
}
resetPaths();


/*
// move to package.json
const babelOptions = {
  optional: ['es7.objectRestSpread'],
  plugins: ['object-assign'],
};*/

const autoprefixerOptions = {
  //FIXME Autoprefixer会分析CSS代码，并且根据Can I Use所提供的资料来决定要加上哪些浏览器前缀
  //browsers: ['> 1%', 'last 2 versions', 'ie 10']
  browsers : [
    'ie >= 9',
    'ie_mob >= 10',
    'ff >= 30',
    'chrome >= 34',
    'safari >= 7',
    'opera >= 23',
    'ios >= 7',
    'android >= 2.3',
    'bb >= 10'
  ]
};

const replaceVersion = function() {
  return $.replace('__VERSION__', pkg.version);
};

const addBanner = function() {
  return $.header(banner);
};

const buildBanner = function() {
  return $.header(`/*! Updated: ${$.util.date(Date.now(), 'isoDateTime')} */`);
};

gulp.task('clean', () => {
  return del(['dist', 'www', 'lib']);
});

gulp.task('clean:app', () => {
  console.log('Dir: ' + appDir);
  return del([appPaths.appDist]);
});

/**
 * Build Amaze UI Touch
 */

gulp.task('build:clean', () => {
  return del([paths.dist, 'lib']);
});

gulp.task('style:scss', () => {
  var options = appDir === 'docs/_app' ? paths : appPaths;
  //if(!options.style) return;
  return gulp.src(options.style)
    .pipe($.sass({
      outputStyle: 'expanded'
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer(autoprefixerOptions))
    .pipe(addBanner())
    .pipe(gulp.dest(options.dist + '/css'))
    .pipe($.if(!isProduction, gulp.dest(appPaths.appDist + '/css')))
    .pipe($.csso())
    .pipe(addBanner())
    .pipe($.rename({suffix: '.min'}))
    .pipe(gulp.dest(options.dist + '/css'))
    .pipe($.if(isProduction, gulp.dest(appPaths.appDist + '/css')));
});

gulp.task('style:fonts', () => {
  return gulp.src(paths.fonts)
    // .pipe(gulp.dest(paths.dist + '/css/fonts'))
    .pipe(gulp.dest(appPaths.appDist + '/css/fonts'));
});

gulp.task('copy:imgs', () => {
  return gulp.src(appPaths.imgs)
    // .pipe(gulp.dest(paths.dist + '/i'))
    .pipe(gulp.dest(appPaths.appDist + '/i'));
});

gulp.task('style:watch', () => {
  gulp.watch(paths.scssModules, ['style:scss']);
});

gulp.task('style', ['style:scss', 'style:fonts', 'copy:imgs']);
gulp.task('styleDev', ['style:scss', 'style:fonts', 'copy:imgs', 'style:watch']);

// transform ES6 & JSX
gulp.task('build:babel', () => {
  return gulp.src('src/js/**/*')
    .pipe(replaceVersion())
    .pipe($.babel())
    .pipe(gulp.dest('lib'));
});

// 由于 browserify-shim 打包 UMD 不会自动添加 CommonJS、AMD 依赖，
// 所以使用 webpack 打包。考虑将 browserify 的工作都转移到 webpack 以减少 npm 依赖数量。
gulp.task('build:pack', () => {
  return gulp.src(paths.jsEntry)
    // .pipe(named())
    .pipe(webpack(assign({},webpackConfig,{
      debug: !isProduction,
      entry: {
        //分开打包，需要指定文件分组，并且配置resolve解释依赖文件的位置
        'index': paths.jsEntry,
        'vendor': ['react', 'react-dom', 'redux', 'react-redux', 'redux-thunk', 'classnames'],  //这里可以结合智能提取公共脚本插件
      },
      output: {
        // library: 'AMUITouch',
        libraryTarget: 'umd',
        //path: path.join(__dirname, 'public/js/dest'),
        filename: '[name].bundle.js'
      },
      resolve: {
        extensions: ["", ".js", ".jsx", '.es6'],
        // 这里 root 指谁的跟，项目根目录？可以配置 path.join(__dirname, '/src/js') 吗
        root: path.join(__dirname, '/'),
        modulesDirectories: ["node_modules"]
      }
    })))
    .pipe(replaceVersion())
    .pipe(addBanner())
    // .pipe($.rename('amazeui.touch.js'))
    .pipe(gulp.dest(paths.dist))
    .pipe($.uglify())
    .pipe(addBanner())
    .pipe($.rename({suffix: '.min'}))
    .pipe(gulp.dest(paths.dist));
});

gulp.task('build', (callback) => {
  runSequence(
    'build:clean',
    ['style', 'build:babel', 'build:pack',],
    callback
  );
});

gulp.task('html:replace', () => {
  const rFrom = '__ENV__';
  const rTo = isProduction ? '.min' : '';
  const replaceEnv = function(options) {
    return gulp.src(options.src)
      .pipe($.replace(rFrom, rTo))
      .pipe(gulp.dest(options.dist));
  };
  let app = replaceEnv({
    src: appPaths.appIndex,
    dist: appPaths.appDist,
  });

  return merge(app);
});

let bsf = (options) => {
  const babelify = ['babelify'];
  let transform = Array.isArray(options.transform) ?
    babelify.concat(options.transform) : babelify;

  return watchify(browserify({
    cache: {},
    packageCache: {},
    entries: options.entries,
    debug: !isProduction,
    transform: transform,
    // path map for fake `amazeui-touch` in `./kitchen-sink/`
    // @see https://github.com/vigetlabs/gulp-starter/issues/17
    paths: ['./'+ appDir +'/'],
  }));
};

let bundler = (options) => {
  let stream = (
    options.b.bundle()
      .on('error', $.util.log.bind($.util, 'Browserify Error'))
      .pipe(source('app.js'))
      .pipe(buffer())
      .pipe(replaceVersion())
      .pipe($.if(!isProduction, gulp.dest(options.dist)))
      .pipe($.size({
        title: `[${options.title}]`,
        showFiles: true,
      }))
  );

  return !isProduction ? stream : stream.pipe($.uglify())
    .pipe($.rename({suffix: '.min'}))
    .pipe(buildBanner())
    .pipe(gulp.dest(options.dist))
    .pipe($.size({
      showFiles: true,
      title: `[${options.title}] - minified`,
    }))
    .pipe($.size({
      showFiles: true,
      gzip: true,
      title: `[${options.title}] - gzipped`,
    }));
};

// gulp.task('docs:js', () => {
//   const docBundle = bsf({
//     entries: [appPaths.js],
//     transform: [[markedify, {marked: getMarked()}], 'brfs']
//   });
//
//   const docsBundleOptions = {
//     title: 'Docs',
//     b: docBundle,
//     dist: appPaths.dist,
//   };
//
//   docBundle.on('update', bundler.bind(null, docsBundleOptions))
//     .on('log', $.util.log);
//
//   return bundler(docsBundleOptions);
// });

// kitchen-sink
// APP
gulp.task('app:build', () => {
  var appBundle = bsf({
    // vendors: [
    //   "react",
    //   "react-dom",
    //   "react-addons-css-transition-group",
    //   "react-router"
    // ],
    entries: appPaths.appEntry,
    // output: {
    //   path: 'www/app/js'
    // }
  });
  var appOptions = {
    title: 'App',
    b: appBundle,
    dist: appPaths.appDist + '/js',
  };

  appBundle.on('update', bundler.bind(null, appOptions))
    .on('log', $.util.log);

  return bundler(appOptions);
  //return bundler(docsBundleOptions);
});

gulp.task('docs:style', () => {
  let stream = gulp.src(appPaths.style)
    .pipe($.sass({
      outputStyle: 'expanded'
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer(autoprefixerOptions));

  return !isProduction ? stream.pipe(gulp.dest(appPaths.dist)) :
    stream.pipe($.csso())
      .pipe(buildBanner())
      .pipe($.rename({suffix: '.min'}))
      .pipe(gulp.dest(appPaths.dist));
});

gulp.task('docs:replace', () => {
  const rFrom = '__ENV__';
  const rTo = isProduction ? '.min' : '';
  const replaceEnv = function(options) {
    return gulp.src(options.src)
      .pipe($.replace(rFrom, rTo))
      // replace stat code on dev
      .pipe($.replace(/<!--STAT_CODE_START-->(.+)<!--STAT_CODE_END-->/g,
        (match, $1) => {
        return isProduction ? $1 : '';
      }))
      .pipe(gulp.dest(options.dist));
  };
  let docs = replaceEnv({
    src: `${docsDir}/index.html`,
    dist: appPaths.dist,
  });
  let ks = replaceEnv({
    src: appPaths.appIndex,
    dist: appPaths.appDist,
  });

  return merge(docs, ks);
});

gulp.task('watch', () => {
  // gulp.watch( appDir + '**/*.js', ['app:webpack']);
  gulp.watch( appDir + '/*.html', ['html:replace']);
  gulp.watch( appDir + '/style/*.scss', ['styleDev']);
});

gulp.task('watch:webpack', () => {
  // gulp.watch( appDir + '**/*.js', ['app:webpack']);
  gulp.watch( appDir + '/js/**/*.js', ['app:webpack']);
});

gulp.task('app:server', ['watch'], () => {
  let bs = BS.create();
  bs.init({
    server: [appPaths.appDist],
    open: 'external',
  });

  gulp.watch(`${appPaths.dist}/**/*`, bs.reload);
  gulp.watch(`${appPaths.styleDir}/*`, ['docs:style']);
});

gulp.task('docs', (callback) => {
  resetPaths('docs/_app');
  // runSequence('clean',
  //   ['styleDev', 'docs:style', 'docs:js', 'docs:replace', 'app:build'],
  //   'docs:server',
  //   callback);
  runSequence(
      'clean:app',
      ['styleDev', 'html:replace', 'build', 'app:build'],
      'app:server',
      callback);
});

gulp.task('default', ['ks']);

gulp.task('ks', (callback) => {
  resetPaths('kitchen-sink');

  runSequence(
    'clean:app',
    'quote:styles',
    ['styleDev', 'html:replace', 'app:build'],
    'watch:quote:styles',
    'app:server',
    callback);
});

gulp.task('app', (callback) => {
  resetPaths('app');

  runSequence(
    'clean:app',
    ['styleDev', 'style', 'html:replace', 'app:build'],
    'app:server',
    callback);
});

gulp.task('dev', (callback) => {
  resetPaths('_app');

  runSequence(
    'clean:app',
    ['styleDev', 'html:replace', 'app:build'],
    // ['styleDev', 'html:replace', 'app:webpack'],
    'app:server',
    callback);
});


/**

npm install history humps isomorphic-fetch lodash normalizr react react-dom react-redux react-router redux redux-logger redux-router redux-thunk --save

npm install babel babel-core babel-loader concurrently babel-plugin-react-transform react-transform-hmr webpack webpack-dev-middleware webpack-hot-middleware --save-dev

npm install react-tools redux-devtools redux-devtools-dock-monitor redux-devtools-log-monitor --save-dev
*/

import startWebpackConfig from './start/webpack.config';
// import webpack2 from 'webpack';
// run webpack
gulp.task('app:webpack', (callback) => {
  return gulp.src(appPaths.appEntry)

    .pipe(webpack(assign({}, startWebpackConfig, {  //startWebpackConfig
      debug: true,//!isProduction,
      entry: {
        // watch: true,
        devtool: 'source-map', //生成sourcemap,便于开发调试
        // 'path': path.join(__dirname, './start/js'),
        //分开打包，需要指定文件分组，并且配置resolve解释依赖文件的位置
        'index': './' + appPaths.appEntry,//'index.js',//appPaths.appEntry,
        'vendor': ['react', 'react-dom', 'redux', 'react-redux', 'redux-thunk', 'classnames'], //这里可以结合智能提取公共脚本插件
      },
      output: {
        filename: '[name].bundle.js'
      },
    })))
    .pipe($.sourcemaps.init({loadMaps: true}))
    // .pipe(replaceVersion())
    // .pipe(addBanner())
    // .pipe($.rename('app.js'))
    .pipe(through.obj(function (file, enc, cb) {
      // Dont pipe through any source map files as it will be handled
      // by gulp-sourcemaps
      var isSourceMap = /\.map$/.test(file.path);
      if (!isSourceMap) this.push(file);
      cb();
    }))
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest(appPaths.appDist + '/js'))
    .pipe($.uglify())
    // .pipe(addBanner())
    .pipe($.rename({suffix: '.min'}))
    .pipe(gulp.dest(appPaths.appDist + '/js'));

});

gulp.task('start', (callback) => {
  resetPaths('start', 'index.js');

  runSequence(
    'clean:app',
    'quote:styles',
    // ['styleDev', 'html:replace', 'app:webpack'],
    ['styleDev', 'html:replace', 'app:build'],
    // 'watch:webpack',
    // 'watch:quote',
    'app:server',
    callback);
});







// 编译 SCSS，添加浏览器前缀
// 要配置 autoprefixerOptions
gulp.task('quote:styles', function () {
  var s = (
    gulp.src(appPaths.quoteStyles)
    .pipe($.rename('smacss.css'))
    .pipe($.sourcemaps.init())
    //.pipe($.plumber())  //自动处理全部错误信息防止因为错误而导致 watch 不正常工作
    .pipe($.sass())
    .pipe($.autoprefixer(autoprefixerOptions))
    .pipe($.sourcemaps.write())

    .pipe(gulp.dest(appPaths.appDist + '/css'))
  );
  return !isProduction ? s : s.pipe($.csso())
    .pipe($.rename({suffix: '.min'}))
    .pipe(md5(10, appPaths.quotePage))
    .pipe(gulp.dest(appPaths.appDist + '/css'))
    .pipe($.size({title: 'styles'}));
});
gulp.task('watch:quote:styles', () => {
  // gulp.watch( appDir + '**/*.js', ['app:webpack']);
  gulp.watch( appPaths.quoteSrc + '/styles/**/*.scss', ['quote:styles']);
});


/**
// 上传七牛cdn
var qn = require('gulp-qn');
var qiniu = {
  accessKey: '6sBCo463jJOCnBIYX__uy9avZ7C2hj_MHb-ffKAr',
  secretKey: '3vPk7fB0HcwL5V9E2AErHuR19HM389eYqdvQcncL',
  bucket: 'xdemo',
  domain: 'http://7xik9a.com1.z0.glb.clouddn.com'
};

//某任务中插入即可
.pipe(qn({
  qiniu: qiniu,
  prefix: 'gmap'
}))


// 可以在后面输入环境参数 如： gulp build -e production
// 参看 http://keenwon.com/1496.html 以及 https://github.com/keenwon/fede2
var argv = require('optimist').argv,
*/
